[WIP] Setup WrappedRule and SubRules to combine rules

Mauin opened the pull request:
This is a first attempt to discuss about the general idea. Consider it WIP. I mainly want to discuss about this approach and if it makes sense at all.

The `DetektVisitor` gets the basic `visit` responsibility and `Context`. That means that `RuleSet` can then iterate over `DetektVisitor` so that the combined rule doesn't need to implement `Issue`.

Wrapping rules should then implement `DetektVisitor` instead of `Rule`. The wrapping rules can then implement multiple `SubRule`s which can easily be invoked with the extension methods I've added.

I have migrated the `MaxLineLength` rule to this style. It is visiting the file, splitting it's content into each line and then passing down the `Sequence<String>` to the actual `MaxLineLength` rule. If we now had a 2nd rule that also needed this data we could write the wrapping rule as: 

```
class CombinedRules(val config: Config = Config.empty) : DetektVisitor() {

    override fun visitKtFile(file: KtFile) {
        val lines = file.text.splitToSequence("\n")
       lines.applyRules { listOf(
           MaxLineLengthRule(config, file)
           TodoRule(config, file)
        )}.forEach { report(it) }
    }
}
```

Not yet really happy about the having to re-report all found issues to the combined rule. That's definitely one point I'd love to improve on. But first I want to get some feedback.

arturbosch commented:
I like the concept. 

Yes, the sub rules should report their findings to their context and the CombinedRules should override the getFinsings method and report all aggregated findings.

What I do not like is that a RuleSet now operates on DetektVisitor's. It's the naming I find odd. I also use the DetektVisitor's to collect some metrics which also do not need the extra visit methods and is active check. Maybe an extra RuleBase class would be better?

Mauin commented:
Thanks for the feedback, sounds good. I'll continue on this later today. 👍

Mauin commented:
Updated the PR and now moved "merging the findings" into the extension function. 

```
lines.reportFindings(context) {
    listOf(MaxLineLengthRule(config, file))
}
```

By passing the context to the `reportFindings` method we now report findings directly to the parent rule.

Still some Tests failing and some cleanup to be done. Also the `MaxLineLengthSpec` should now probably operate on the `SubRule` instead of the `MultiRule`. I'll add some testing support later after I'm sure we're happy with the API.
Feedback welcome!

Mauin commented:
Linking this to #188

arturbosch commented on detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/Rule.kt:
```diff
@@ -1,59 +1,23 @@
-	 */
-	open fun visit(root: KtFile) {
-		ifRuleActive {
-			if (!root.isSuppressedBy(id)) {
```
Where did this check disappear? It saves us from visiting the whole file if there is a file annotation.

Mauin replied:
http://maven.int.s-cloud.net/content/groups/soundcloud-proxy

Mauin replied:
Oops that was a mistake on my part. I'll add it to `Rule#visitCondition()` again.

arturbosch commented on detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/MultiRule.kt:
```diff
@@ -0,0 +1,11 @@
+
+inline fun <T> T.reportFindings(context: Context, rules: () -> List<SubRule<T>>) {
+	val findings = rules.invoke()
+			.map { it.findings }
```
Just use flatMap here, save one List creation :)

arturbosch commented on detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/MultiRule.kt:
```diff
@@ -0,0 +1,11 @@
+import io.gitlab.arturbosch.detekt.api.Context
+
+inline fun <T> T.reportFindings(context: Context, rules: () -> List<SubRule<T>>) {
+	val findings = rules.invoke()
```
invoke is just the creation of the sub rules? Where is apply called? `map { it.findings }` will always be empty?

Mauin commented:
Updated the PR again. I still have to take a look at the tests to migrate the `MaxLineLengthSpec` to not use `MaxLineLength()` which is now a `MultiRule` but the actual `MaxLineLengthRule` now. Otherwise the tests will start failing once we add more rules into this specific `MultiRule`. Thanks for all the feedback so far @arturbosch

arturbosch commented on detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/RuleExtensions.kt:
```diff
@@ -0,0 +1,10 @@
+
+inline fun <T> T.reportFindings(context: Context, rules: () -> List<SubRule<T>>) {
+	rules.invoke()
+			.onEach { it.apply(this) }
```
As we call apply now for all sub rules the isActive check is never done for them. Maybe this check should be done by the MultiRule?

arturbosch commented on detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/MultiRule.kt:
```diff
@@ -0,0 +1,7 @@
+import org.jetbrains.kotlin.psi.KtFile
+
+open class MultiRule : BaseRule() {
+	override fun visitCondition(root: KtFile) = true
```
A Multirule must somehow do the `isActive` check for the sub rules and only `apply` on the active ones. If I see it correctly the MultiRule is run everytime on all sub rules.

arturbosch commented:
Np, thx for the effort on this non trivial api change :+1:

Mauin commented:
Just to update: After a long weekend off I picked this one up again yesterday. Still trying to find some nice way to test the `SubRule` but hoping to get this one done this week 🙃

Mauin commented:
Pushed some updates to this. For now I decided to leave the `isActive` check in the `reportFindings` method. I did this because I would prefer to not heavily bind the `SubRules` to the `MultiRule`. That would enforce that all `SubRule`s in a `MultiRule` work on the same type (`List<String>` in the Rule of this PR).
Without binding them together it would be possible to write a `MultiRule` that has a `SubRule` that parses the entire file text (`SubRule<String>`), but also checks each line of the file (`SubRule<List<String>>`).

Otherwise if we decide to bind them together, we could add a `abstract fun rules(): List<SubRule<T>>` to the `MultiRule` which provides all the rules and can then also perform the `isActive` check. 
But as of now the `MultiRule` does not know about the `SubRule`s it is handling.

arturbosch commented:
I did some enhancements for the sub and multi rules, so they do not need an extra ktFile attribute and the `isActive` check is used like in all other rules.

```kotlin
abstract class SubRule<in Element>(config: Config) : Rule(config) {

	abstract fun apply(element: Element)
}
```

```kotlin
inline fun <T> T.reportFindings(context: Context, rules: () -> List<SubRule<T>>) {
	rules.invoke() // No isActive here
			.onEach { it.apply(this) }
			.flatMap { it.findings }
			.apply { context.report(this) }
}
```

The active check is done in `visitCondition`. We do not splitToSequence anymore on each file.
```kotlin
class FileParsingRule(val config: Config = Config.empty) : MultiRule() {

	private val rules = listOf(MaxLineLength(config))
	private val isActive = rules.find { it.active } != null

	override fun visitCondition(root: KtFile) = isActive

	override fun visitKtFile(file: KtFile) {
		val lines = file.text.splitToSequence("\n")
		val fileContent = KtFileContent(file, lines)
		fileContent.reportFindings(context) { rules }
	}

}
```

To remove the extra KtFile attribute I used a simple data class

```kotlin
data class KtFileContent(val file: KtFile, val content: Sequence<String>)
```

The MaxLineRule slightly changed

```kotlin
class MaxLineLength(config: Config = Config.empty) : SubRule<KtFileContent>(config) {

	override val issue = Issue(javaClass.simpleName,
			Severity.Style,
			"Line detected that is longer than the defined maximum line length in the code style.",
			Debt.FIVE_MINS)

	private val maxLineLength: Int
			= valueOrDefault(MaxLineLength.MAX_LINE_LENGTH, MaxLineLength.DEFAULT_IDEA_LINE_LENGTH)
	private val excludePackageStatements: Boolean
			= valueOrDefault(MaxLineLength.EXCLUDE_PACKAGE_STATEMENTS, MaxLineLength.DEFAULT_VALUE_PACKAGE_EXCLUDE)
	private val excludeImportStatements: Boolean
			= valueOrDefault(MaxLineLength.EXCLUDE_IMPORT_STATEMENTS, MaxLineLength.DEFAULT_VALUE_IMPORTS_EXCLUDE)

	override fun apply(element: KtFileContent) {
		var offset = 0
		val lines = element.content
		val file = element.file
		lines.filter { filterPackageStatements(it) }
				.filter { filterImportStatements(it) }
				.map { it.length }
				.forEach {
					offset += it
					if (it > maxLineLength) {
						report(CodeSmell(issue, Entity.from(file, offset)))
					}
				}
	}

```

arturbosch commented:
I did not quite understand what you meant with binding the rules (same T ?) I think it is still possible to have a sub rule on `ktFileContent.file.text` and the other on `ktFileContent.lines`

I do not see the usecase for now. The multirule should always know its subrules or?

arturbosch commented:
What do you think of my suggestions?

Mauin commented:
@arturbosch I'll work on integrating them later today. Thanks for the suggestions. Really like getting the `KtFile` out of the constructor but rather passing it to the `apply` method.

arturbosch commented:
Oh, the isActive check in `reportFindings` is of course necessary!

```kotlin
inline fun <T> T.reportFindings(context: Context, rules: () -> List<SubRule<T>>) {
	rules.invoke() // No isActive here
			.onEach { it.apply(this) }
			.flatMap { it.findings }
			.apply { context.report(this) }
}
```

Mauin commented:
The `isActive` check would still need a reference of the file. As an idea I've extracted `File` abstract class and only allow the `reportFindings` method on classes extending it. 

Otherwise I think the changes sound good.
